Implementation for Joint Shapley Values

RBorn02 opened the pull request:
Runs 7 out of 10 Tests if we comment out trained_model = None in cough_and_fever and cough_and_fever_10_90. Times out on mnist and StressNbFeatures due to large number of features.

Karim-53 reviewed (commented):
- [x] rename file to `joint_shapley.py`
- [ ] add it to `explainers/__init__.py`
- [ ] add it `valid_explainers`

Karim-53 commented on explainers/joint_shapley.py:
```diff
@@ -0,0 +1,212 @@
+        super().__init__()
+        self.trained_model = trained_model
+        if predict_func is None:
+           self.value_f = self.trained_model.predict
+        else:
+            self.value_f = predict_func
```
```suggestion
           predict_func = self.trained_model.predict
        else:
            predict_func = predict_func

        def predict_func_numpy(X):
            if isinstance(X,pd.DataFrame):
                return predict_func(X.values)
            else:
                return predict_func(X)

        self.value_f = predict_func_numpy
```

Karim-53 commented:
I just added a fix. now you can run cough and fever test

- [ ] please reduce the execution time limit (`TIME_LIMIT`) to 60 seconds to not lose a lot of time (no need to commit it)
- [ ] plz add the results generated in `data\02_experiment_output\results.csv` plz commit only the new line
- [ ] there is still one small bug could you please have a look ?
![image](https://user-images.githubusercontent.com/33978275/206310836-2a2a488b-3330-4a8e-aed2-33d2cd50636a.png)

RBorn02 commented:
I fixed the issue with the FoolingPertubationAlg Test and updated the result and explainer csv with joint shapley. I also wrote a small description for each interaction in the archipelago interaction detection test.

Karim-53 commented:
Hi,

the separator in the csv file changed from , to ;
This might break the rest of the deployment
- [ ] Could you please keep the old version? only one line should be added to [explainer.csv](https://github.com/Karim-53/Compare-xAI/pull/10/files#diff-035dc65d9525c9cf121efce0dde94f9f820dbcdd0f67cfea102b6be5fc733116)
- [ ] same for test.csv
You can open the csv with Pycharm and install the following plugin 
![image](https://user-images.githubusercontent.com/33978275/206671558-fa6db16d-bd61-4dc5-946e-89670f9a4c33.png)
by the way did you start using PyCharm ?

Karim-53 commented:
- [ ] plz add the following line to paper.csv
harris2021joint,0.5,JointShapley,,,"@article{harris2021joint,
  title={Joint shapley values: a measure of joint feature importance},
  author={Harris, Chris and Pymar, Richard and Rowat, Colin},
  journal={arXiv preprint arXiv:2107.11357},
  year={2021}
}",

Karim-53 commented:
- [ ] no need to change `data/03_experiment_output_aggregated/explainer.csv`
normally, this file is modified by the script `src/aggregate_data.py` but no need to run that for now
